src/t_zset.c

```diff
@@ -2184,6 +2188,165 @@ inline static void zunionInterAggregate(double *target, double val, int aggregat
+    sds tmp;
+
+    /* With algorithm 1 it is better to order the sets to subtract
+     * by decreasing size, so that we are more likely to find
+     * duplicated elements ASAP. */
+    qsort(src+1,setnum-1,sizeof(zsetopsrc),zuiCompareByRevCardinality);
+
+    memset(&zval, 0, sizeof(zval));
+    zuiInitIterator(&src[0]);
+    while (zuiNext(&src[0],&zval)) {
+        double score, value;
+        int exists = 0;
+
+        score = zval.score;
+        if (isnan(score)) score = 0;
+
+        for (j = 1; j < setnum; j++) {
+            /* It is not safe to access the zset we are
+                * iterating, so explicitly check for equal object. */
+            if (src[j].subject == src[0].subject ||
+                    zuiFind(&src[j],&zval,&value)) {
+                exists = 1;
+                break;
+            }
+        }
+
+        if (!exists) {
+            tmp = zuiNewSdsFromValue(&zval);
+            znode = zslInsert(dstzset->zsl,score,tmp);
+            dictAdd(dstzset->dict,tmp,&znode->score);
+            if (sdslen(tmp) > *maxelelen) *maxelelen = sdslen(tmp);
+        }
+    }
+}
+
+
+void zdiffAlgorithm2(zsetopsrc *src, long setnum, zset *dstzset, size_t *maxelelen) {
+    /* DIFF Algorithm 2:
+     *
+     * Add all the elements of the first set to the auxiliary set.
+     * Then remove all the elements of all the next sets from it.
+     *
+     * This is O(N) where N is the sum of all the elements in every
+     * set. */
+    int j;
+    int cardinality = 0;
+    zsetopval zval;
+    zskiplistNode *znode;
+    sds tmp;
+    double score;
+    dictEntry *de;
+
+    for (j = 0; j < setnum; j++) {
+        if (zuiLength(&src[j]) == 0) continue;
+
+        memset(&zval, 0, sizeof(zval));
+        zuiInitIterator(&src[j]);
+        while (zuiNext(&src[j],&zval)) {
+            if (j == 0) {
+                tmp = zuiNewSdsFromValue(&zval);
+                znode = zslInsert(dstzset->zsl,zval.score,tmp);
+                dictAdd(dstzset->dict,tmp,&znode->score);
+                cardinality++;
+            } else {
+                tmp = zuiNewSdsFromValue(&zval);
+                de = dictUnlink(dstzset->dict,tmp);
+                if (de != NULL) {
+                    /* Get the score in order to delete from the skiplist later. */
+                    score = *(double*)dictGetVal(de);
+
+                    /* Delete from the hash table and later from the skiplist.
+                    * Note that the order is important: deleting from the skiplist
+                    * actually releases the SDS string representing the element,
+                    * which is shared between the skiplist and the hash table, so
+                    * we need to delete from the skiplist as the final step. */
+                    dictFreeUnlinkedEntry(dstzset->dict,de);
+
+                    /* Delete from skiplist. */
+                    int retval = zslDelete(dstzset->zsl,score,tmp,NULL);
+                    serverAssert(retval);
```

Comment:
i agree with your suggestion, let's move this into a common function that serves both this one and zsetDel, and returns a value so the caller knows if the entry was deleted and can carry out the other tasks forthat `if`

Reply:
Done. I'm not sure the name of the new function is really appropriate though, could use some help with it. I called it `zsetslDel`, seeing as it is a function to delete an element from a zset encoded as a skiplist (plus dict).

Reply:
i don't have any definite answer. i think i would like to avoid creating a new type of prefix.
i would suggest to just use a `zset` prefix, and good top comment.
but `zsetDel` is taken 8-).
maybe call it `zsetRemoveFromSkiplist` ?

Reply:
Done!